contrib/epee/include/net/net_utils_base.h

```diff
@@ -107,6 +108,73 @@ namespace net_utils
 	inline bool operator>=(const ipv4_network_address& lhs, const ipv4_network_address& rhs) noexcept
 	{ return !lhs.less(rhs); }
 
+	class ipv6_network_address
+	{
+	protected:
+		boost::asio::ip::address_v6 m_address;
+		std::string m_address_str;
+		uint16_t m_port;
+
+	public:
+		ipv6_network_address()
+			: ipv6_network_address("::1", 0)
+		{}
+
+		ipv6_network_address(const std::string& ip, uint16_t port)
+			: m_address(boost::asio::ip::address_v6::from_string(ip)), m_address_str(ip), m_port(port)
+		{
+			m_address_str = m_address.to_string();
+		}
+
+		bool equal(const ipv6_network_address& other) const noexcept;
+		bool less(const ipv6_network_address& other) const noexcept;
+		bool is_same_host(const ipv6_network_address& other) const noexcept
+		{ return m_address == other.m_address; }
+
+		std::string ip() const noexcept { return m_address_str; }
+		uint16_t port() const noexcept { return m_port; }
+		std::string str() const;
+		std::string host_str() const;
+		bool is_loopback() const;
+		bool is_local() const;
+		static constexpr address_type get_type_id() noexcept { return address_type::ipv6; }
+		static constexpr zone get_zone() noexcept { return zone::public_; }
+		static constexpr bool is_blockable() noexcept { return true; }
+
+		template <bool is_store, typename std::enable_if<!is_store>::type* = nullptr>
+		static void post_deserialize(ipv6_network_address& instance)
+		{
+			instance.m_address = boost::asio::ip::address_v6::from_string(instance.m_address_str);
+			instance.m_address_str = instance.m_address.to_string();
+		}
+
+		template <bool is_store, typename std::enable_if<is_store>::type* = nullptr>
+		static void post_deserialize(const ipv6_network_address& instance)
+		{
+		}
+
+		static const uint8_t ID = 2;
+		BEGIN_KV_SERIALIZE_MAP()
+			KV_SERIALIZE(m_address_str)
+			KV_SERIALIZE(m_port)
+
+			post_deserialize<is_store>(this_ref);
+		END_KV_SERIALIZE_MAP()
+	};
```

Comment:
That all seem dangerous. Is it possible to just remove that m_address_str, and just have the ipv6 var ? host_str calls to_string. No need for this post_deserialize.

Reply:
Would it be preferable to use boost's address_v6 to make the string address canonical in the case where the address is passed to the constructor and trust that if we're deserializing one it is canonical and that address_v6.to_string() does indeed return at least a consistent (if not canonical, consistent should be fine) string, or to store the address bytes?  I can use the underlying representation address_v6 uses and that should work nicely, but I'll have to add std::array to the list of stl containers supported by KV_SERIALIZE.

Reply:
I'm not sure I understand that :)

- having a string seems wrong since several strings can represent the same address
- I assume the boost address_v6 is just 8 bytes and generally behaves sensibly; if not you can just make your own
- For serialization, you can save as a blob assuming the boost object is not a complex thing, which it has no reason to be, but I'm guessing here

Reply:
re: strings, I agree that that's not the best way to do things, though as I said if address_v6::to_string() is consistent in how it represents an address then there shouldn't be an issue of multiple strings for the same address.

address_v6 uses a std::array of 16 bytes internally.  It does behave sensibly but again I'd need to add it to the list of containers supported by KV_SERIALIZE.  This seems like the best approach, I just thought I'd confirm as much.

The problem with serializing as a different representation (e.g. blob it up before serializing) is that on deserialize the underlying field must be populated, hence the post_serialize above.  I agree that this approach feels clumsy; at the time I hadn't realized that address_v6 exposes the raw bytes, simply misread the header.

I'll go ahead and use the std::array as the underlying representation as it seems like the cleanest approach at this point.

Reply:
Can you not use KV_SERIALIZE_POD_AS_BLOB ?

Reply:
Ok, nevermind, KV_SERIALIZE did not want to handle std::array and if it's not going to be that simple to tell it to it's probably not worth it.

So the issue at hand is that doing a "post-serialize" function is a bit messy.  I agree with that, but the alternative seems to be running code every time one of the access functions is used, either to check if the struct has been updated since being deserialized or to do what that updating would do every time.

That seems worse, so I'd say it's between post-serialize or trust that address_v6::to_string() will always give the same string for a given address.  It's possible I'm missing something obvious as I am rather quite tired; I'm genuinely not trying to be obtuse.  Given that serializing std::array wasn't a quick win and post-serialize is arguably ugly, I'm inclined to say trust that address_v6::to_string() is consistent.

Reply:
Should have hit refresh.  Will take a look at KV_SERIALIZE_POD_AS_BLOB, not sure if that will sort it.

Reply:
https://paste.debian.net/hidden/f4a22592/ are the changes I'd make. This uses only a address_v6 object. I think there might even be fewer conversions overall. I've only built it, not tried it.

Reply:
Thanks for that.  There were a couple issues to touch up with that patch but no worries.  Biggest is that address_v6.from_string is a static member function.  Why the compiler allows instance.static_function() is beyond me, but finding that sure was fun.

Reply:
That const_cast should not be needed, but actually was when I tried it. Not sure why... I might look at that later too.